internal/service/cluster/discovery/join_peers.go

```diff
@@ -0,0 +1,178 @@
+			// It is still useful to know if user provided an address that we could not resolve, even
+			// if another addresses resolve successfully, and we don't return an error. To keep things simple, we're
+			// not including more detail as it's available through debug level.
+			level.Warn(opts.Logger).Log("msg", "failed to resolve provided join address", "addr", addr)
+		}
+	}
+
+	if len(result) == 0 {
+		return nil, fmt.Errorf("failed to find any valid join addresses: %w", deferredErr)
+	}
+	return result, nil
+}
+
+type addressResolver func(addr string) ([]string, error)
+
+func ipResolver(log log.Logger) addressResolver {
+	return func(addr string) ([]string, error) {
+		// Check if it's IP and use it if so.
+		ip := net.ParseIP(addr)
+		if ip == nil {
+			return nil, fmt.Errorf("could not parse as an IP or IP:port address: %q", addr)
+		}
+		level.Debug(log).Log("msg", "found an IP cluster join address", "addr", addr)
+		return []string{ip.String()}, nil
+	}
+}
+
+func dnsAResolver(opts Options, ctx context.Context) addressResolver {
+	// Default to net.LookupIP if not provided. By default, this will look up A/AAAA records.
+	ipLookup := opts.lookupIPFn
+	if ipLookup == nil {
+		ipLookup = net.LookupIP
+	}
+	return dnsResolver(opts, ctx, "A/AAAA", func(addr string) ([]string, error) {
+		ips, err := ipLookup(addr)
+		result := make([]string, 0, len(ips))
+		for _, ip := range ips {
+			result = append(result, ip.String())
+		}
+		return result, err
+	})
+}
+
+func dnsSRVResolver(opts Options, ctx context.Context) addressResolver {
+	// Default to net.LookupSRV if not provided.
+	srvLookup := opts.lookupSRVFn
+	if srvLookup == nil {
+		srvLookup = net.LookupSRV
+	}
+	return dnsResolver(opts, ctx, "SRV", func(addr string) ([]string, error) {
+		_, addresses, err := srvLookup("", "", addr)
+		result := make([]string, 0, len(addresses))
+		for _, a := range addresses {
+			result = append(result, a.Target)
+		}
+		return result, err
+	})
+}
+
+func dnsResolver(opts Options, ctx context.Context, recordType string, dnsLookupFn func(string) ([]string, error)) addressResolver {
+	return func(addr string) ([]string, error) {
+		_, span := opts.Tracer.Tracer("").Start(
+			ctx,
+			"ClusterPeersDNSLookup",
+			trace.WithSpanKind(trace.SpanKindInternal),
+			trace.WithAttributes(attribute.String("addr", addr)),
+			trace.WithAttributes(attribute.String("record_type", recordType)),
+		)
+		defer span.End()
+
+		addresses, err := dnsLookupFn(addr)
+		if err != nil {
+			level.Debug(opts.Logger).Log("msg", "failed to resolve DNS records", "addr", addr, "record_type", recordType, "err", err)
+			span.SetStatus(codes.Error, err.Error())
+			return nil, fmt.Errorf("failed to resolve %q records: %w", recordType, err)
+		}
+
+		// Use all the addresses found
+		level.Debug(opts.Logger).Log("msg", "received DNS query response", "addr", addr, "record_type", recordType, "records_count", len(addresses))
+		var result []string
+		for _, a := range addresses {
```

Comment:
Can we return the addresses without adding to result? kthey are both []string.

Reply:
Yes! leftover from refactoring that I've missed, good catch!